internal/virtualenvironment/virtualenvironment.go

```diff
@@ -78,6 +79,23 @@ func (v *VirtualEnvironment) Setup(installIfNecessary bool) *failures.Failure {
 	return nil
 }
 
+// FindProgramOnPath tries to find the cmd on the PATH environment variable
+func FindProgramOnPath(cmd string, envMap map[string]string) string {
+	var bins []string
+	if p, ok := envMap["PATH"]; ok {
+		bins = strings.Split(p, string(os.PathListSeparator))
+	}
+
+	for _, b := range bins {
+		cmdPath := filepath.Join(b, cmd)
+		if fileutils.IsExecutable(cmdPath) {
```

Comment:
This only works if the user typed the fully qualified executable name. (eg. "perl.exe" on windows, rather than just "perl").

Why do we need the full path though? We're setting up the environment for them, their shell should be doing this, not us.

Reply:
That is a good point:  So, we would also need to take the `PATHEXT` environment variable into account.

On you question why we need this:
When we start a program in the subshell, we parse our own environment variables and existing `.bashrc`, `.zshrc` etc files in.  And possibly they are merged with an existing environment.  In the tests that created a situation, where `<cachedir>/bin` always appeared early in the `PATH`.  So, when I call 

1. I call `<cachedir>/bin/python3`
1. that calls `state shim -- python3`
1. after running it through `subshell.Run()` we are calling `<cachedir>/bin/python3` again!
1. continue at 1.

This was my suggestion to resolve this issue, but I am now realizing that this is incorrect for two reasons:
1. the `PATHEXT` variable needs to be taken into account on Windows.
1. `state shim` should just ensure that it is not calling itself, maybe by removing the global bin dir from the path.  As this here would fail, if we "shim" an executable that is not in the runtime environment.

Reply:
Also I am realizing:  As I de-activated the `state activate --default` tests, this is not necessary anymore to ensure that the tests are actually working, so how about I revert these changes, and write up a story with what I think needs to be done?